cvejob/version_utils.py

```diff
@@ -0,0 +1,336 @@
+
+
+class ClassifiedVersion(object):
+    """Version classified as affected or unaffected.
+
+    TODO: if Python 3.7: turn into a data class
+    """
+
+    def __init__(self, version, is_affected=False):
+        """Instantiate ClassifiedVersion.
+
+        :param version: str, version
+        :param is_affected: bool, indicator whether the version is affected or not
+        """
+        self.version = version
+        self.is_affected = is_affected
+
+    def __str__(self):
+        return '{version}({status})'.format(version=self.version, status=self.is_affected)
+
+    def __repr__(self):
+        return 'ClassifiedVersion({version}, {status})'.format(
+            version=self.version, status=self.is_affected
+        )
+
+
+def get_configuration_nodes(cve, apps_only=True):
+    """Get configuration nodes from given CVE object.
+
+    :param cve: nvdlib.model.CVE, CVE object
+    :param apps_only: bool, only return configuration nodes affecting application, or all nodes
+    :return: [[str],[str]], list of lists of affected version ranges (as strings)
+    """
+    nodes = list()
+
+    for configuration in cve.configurations.nodes:
+        data = nvdlib.utils.rgetattr(configuration, 'data')
+
+        if apps_only:
+            for node in data:
+                if CPE(node.cpe).is_application():
+                    break
+            else:
+                continue
+
+        nodes.append(nvdlib.utils.rgetattr(configuration, 'data'))
+
+    return nodes
+
+
+def get_version_ranges(nodes):
+    """Get version ranges from configuration nodes.
+
+    :param nodes: list of configuration nodes as returned from get_configuration_nodes()
+    :return list[str], list of version range strings
+    """
+    return nvdlib.utils.rgetattr(nodes, 'version_range') or []
+
+
+class VersionOperator(Enum):
+    """Enum for version range operators.
+
+    Following operators are currently supported: '==', '<=', '>='.
+    """
+
+    EQ = '=='
+    LE = '<='
+    GE = '>='
+
+
+class VersionSpec(object):
+    """Version specification.
+
+    Version bundled with VersionOperator.
+    """
+
+    def __init__(self, version_str, operator_str):
+        """Instantiate VersionSpec."""
+        self.version = version_str
+        self.operator = VersionOperator(operator_str)
+
+    def contains(self, version_str, cmp_class=ComparableVersion):
+        """Check if `version_str` is in the version spec.
+
+        :param version_str: str, version to check
+        :param cmp_class: class used to compare given version against version spec
+        """
+        this_version = cmp_class(self.version)
+        other_version = cmp_class(version_str)
+
+        if self.operator == VersionOperator.EQ:
+            return this_version == other_version
+        elif self.operator == VersionOperator.LE:
+            return other_version <= this_version
+        elif self.operator == VersionOperator.GE:
+            return other_version >= this_version
```

Comment:
is not it better to use map there?

And what to do in case of improper operator?